### PR TITLE
Add additional validator and metrics for datadog

### PR DESCRIPTION
**Pull request by worldtiki**
### :pencil: Description
- Add additional validator for invalid service names and metrics for datadog forwarder
- Move common validators to correct package (and out of haystacKk)

### CODE REVIEW

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java [hunk 1]
```diff
@@ -36,13 +38,18 @@
...
     public DatadogSpansDispatcher(WebClient datadogClient,
                                   ObjectMapper mapper,
-                                  DatadogForwarderConfigProperties properties) {
+                                  DatadogForwarderConfigProperties properties,
+                                  MetersProvider metersProvider) {
         this.datadogClient = datadogClient;
         this.mapper = mapper.writer();
         this.pending = new ArrayBlockingQueue<>(properties.getQueuedMaxSpans());
+        this.successCounter = metersProvider.forwarderCounter("datadog", true);
+        this.failureCounter = metersProvider.forwarderCounter("datadog", false);
```
Comment (ffeltrinelli) [inline src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:52]: I'm surely missing knowledge here, but I don't see anything in the counter's name that makes me understand that this counter's semantics is that of a success or a failure. The name is just "datadog" with value "true" or "false", I was expecting something like "datadog success" or similar?

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java [hunk 2]
```diff
@@ -88,16 +95,23 @@ private void flush(List<Span> spans) {
...
+                        .body(fromValue(it))
+                        .retrieve()
+                        .toBodilessEntity()
+                        .subscribe(); // FIXME: reactive
+
+                successCounter.increment(spans.size());
+            });
+        } catch (Exception e) {
+            failureCounter.increment(spans.size());
+            throw new RuntimeException(e);
```
Comment (ffeltrinelli) [inline src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:111]: Couldn't you just rethrow the caught exception? Maybe catching `RuntimeException` and just re-throwing it?

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java [hunk 3]
```diff
@@ -116,8 +131,9 @@ private void flush(List<Span> spans) {
     public void addSpan(Span span) {
         try {
             this.pending.add(span);
-        } catch (Exception e) {
-            logger.error("operation=addSpan", e);
+        } catch (IllegalStateException e) {
+            failureCounter.increment();
+            logger.error("Unable to add span with id {} to queue. Queue full.", span.id());
```
Comment (ffeltrinelli) [inline src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:136]: You're logging "queue full" as it's for sure the cause of the `IllegalStateException`. Just double checking, is it the only possible cause of such exception?

File: src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package com.expedia.pitchfork.systems.haystack;
+package com.expedia.pitchfork.systems.common;
```
Comment (ffeltrinelli) [inline src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java:1]: Minor: there is a lot of package moving in this pull request, from `haystack` to `common`. Maybe worth mentioning it in the PR's description and the reason why you are doing it?

File: src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java [hunk 5]
```diff
@@ -35,14 +38,15 @@ public void setup() {
 
     @ParameterizedTest
     @MethodSource("timestamps")
-    public void shouldDiscardSpanIfTimestampIsInvalid(boolean rejectNullTimestamps, int maxDrift, Long timestamp, boolean spanIsKept) {
-        zipkin2.Span zipkinSpan = zipkin2.Span.newBuilder()
+    public void shouldDiscardSpanIfTimestampIsInvalid(boolean rejectNullTimestamps, int maxDrift, Long timestamp, boolean spanIsKept, String[] invalidServiceNames) {
+        zipkin2.Span zipkinSpan = Span.newBuilder()
                 .traceId(zipkinTraceId(123L))
                 .id(zipkinSpanId(456L))
+                .localEndpoint(Endpoint.newBuilder().serviceName("unknown").build())
```
Comment (ffeltrinelli) [inline src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java:45]: This test is named "should discard span if **timestamp** is invalid". But you are providing also an invalid **service name**. Does this mean that the test is not only testing the invalid timestamp?

**Comment by ffeltrinelli** [comment]
Just double checking, I'm missing knowledge here. When you add a feature (small but still a feature) to pitchfork like in this case, don't you:

1) Bump version
2) Update a changelog

?

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java (hunk 1 above)
Comment (worldtiki) [reply src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:52]: The boolean is the `success` tag.

The metrics look like this:
`counter.pitchfork.something{forwarder=datadog, success=true}`
or
 `counter.pitchfork.something{forwarder=datadog, success=false}`

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java (hunk 2 above)
Comment (worldtiki) [reply src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:111]: Yeap I can do that.

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java (hunk 3 above)
Comment (worldtiki) [reply src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:136]: Yeap, it is

image

File: src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java (hunk 1 above)
Comment (ffeltrinelli) [reply src/main/java/com/expedia/pitchfork/systems/datadog/forwarder/DatadogSpansDispatcher.java:52]: Perfect thanks!

File: src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java (hunk 4 above)
Comment (worldtiki) [reply src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java:1]: yeap, done

File: src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java (hunk 5 above)
Comment (worldtiki) [reply src/test/java/com/expedia/pitchfork/systems/common/SpanValidatorTest.java:48]: I've split this into 2 separate tests. One for the timestamps and one for the service names.

**Comment by worldtiki** [comment]
> Just double checking, I'm missing knowledge here. When you add a feature (small but still a feature) to pitchfork like in this case, don't you:
> 
> 1. Bump version
> 2. Update a changelog
> 
> ?

The datadog integration is still experimental so I'm not adding it to the changelog because I don't want it to be used (not until we go through a few iterations).

**Review by ffeltrinelli — APPROVED** [review]

**Pull request merged**